util/cass.py

```diff
@@ -576,6 +578,308 @@ def insert_dataset(stream_key, dataset):
     return ret_val
 
 
+@log_timing(log)
+def insert_dataset(stream_key, dataset):
+    """
+    Insert an xray dataset back into CASSANDRA.
+    First we check to see if there is data in the bins, if there is we
+    either overwrite and update the values or fail and notify the user
+    :param stream_key: Stream that we are using for the data insertion
+    :param dataset: xray dataset we are inserting
+    :return:
+    """
+    # capture the current data from the stream metadata table for later processing, then delete it
+    cur_stream_meta = metadata_service_api.get_stream_metadata_record(*(stream_key.as_tuple()))
+    if cur_stream_meta:
+        del_resp = metadata_service_api.delete_stream_metadata_record(*(stream_key.as_tuple()))
+        del_stts = del_resp.get('statusCode')
+        del_msg = del_resp.get('message')
+        if del_stts == "OK" and del_msg.startswith("Successfully deleted"):
+            count = cur_stream_meta.get('count')
+            first = cur_stream_meta.get('first')
+            last = cur_stream_meta.get('last')
+            log.info("Deleted stream_metadata that summarized the partition_metadata records for %s: " +
+            "count %d, first %f, last %f", stream_key.as_refdes(), count, first, last)
+        else:
+            error_msg = "Failed to delete stream_metadata for %s", stream_key.as_refdes()
+            log.error(error_msg)
+            return error_msg
+
+    # capture current partition metadata for each unique bin in the dataset for later processing
+    cur_part_meta = {}
+    for bin_val in numpy.unique(dataset['bin'].values).tolist():
+        part_meta = metadata_service_api.get_partition_metadata_record(
+            *(stream_key.as_tuple() + (bin_val, CASS_LOCATION_NAME))
+        )
+        if part_meta:
+            # capture the parts that are useful for later
+            cur_part_meta[bin_val] = (part_meta.get('count'), part_meta.get('first'), part_meta.get('last'))
+
+    # Delete the current data from the partition metadata table for the involved bins
+    for pkey, pvals in cur_part_meta.items():
```

Comment:
Would it be possible to update (or delete and re-insert) the partition_metadata on a per bin basis after the particles are inserted into Cassandra? This would minimize the inconsistent state that would result if the process went down after deleting the partition_metadata but before the insertion of the particles and reinsertion of the partition_metadata. Same for stream_metadata.

Reply:
Good point. I will fix that tomorrow.